1/4 : Browser support

dbousque opened the pull request:
Hi there !

This PR brings browser support. I am currently porting [jaeger-node](https://github.com/jaegertracing/jaeger-client-node) to the browser. As part of this effort, I'd like thriftrw to be browser compatible as well.

There aren't a lot of incompatibilities. In fact, the library code per se works well without any change. The only thing that is missing is file system access (see the added documentation for ways around that). Issues when bundling only arise with dependencies of thriftrw-node, which explains the few tricks used in `browserify.js`.

I'v tried to make as few changes as possible, and none were needed for the library, only the tests have been adapted a bit, so that they could run in a browser.

As far as the release model is concerned, I chose to keep both versions in the same package, leveraging the `browser` field in `package.json`, which allows bundle tools to use alternative package versions destined for browsers. If you prefer, we could have a separate package.

Let me know if you'd like more details or require some changes !

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thriftrw/thriftrw-node?pullRequest=165) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thriftrw/thriftrw-node?pullRequest=165) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thriftrw/thriftrw-node?pullRequest=165) it.</sub>

kriskowal commented on browser/karma_start.js:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+__karma__.start = function() {};
```
Missing a newline at end of file here.

dbousque replied:
done

kriskowal commented on browser/browserify.js:
```diff
@@ -0,0 +1,112 @@
+
+// This hack can be dropped if hexer is made a required dependency of bufrw, as discussed here :
+// https://github.com/uber/bufrw/pull/61 and here : https://github.com/uber/bufrw/issues/65
+// All that is needed is a new release to be published.
```
I’ve just published bufrw 0.3 to address this issue.

dbousque replied:
done

dbousque commented:
Thanks for the feedback and the bufrw release @kriskowal. I am not familiar with the thrift protocol. If I understand correctly, you'd like to give the ability to read sources asynchronously, given that users provide an asynchronous method for reading sources in the `fs` object (that they would implement, and so they could make HTTP requests for example). Is that correct ?

kriskowal commented:
> Thanks for the feedback and the bufrw release @kriskowal. I am not familiar with the thrift protocol. If I understand correctly, you'd like to give the ability to read sources asynchronously, given that users provide an asynchronous method for reading sources in the `fs` object (that they would implement, and so they could make HTTP requests for example). Is that correct ?

That is correct, though getting the API right may require experimentation. Ideally:

* There is a common API regardless of whether it’s backed by Node.js or a web fetch.
* All Thrift models share a common memo table and that table can be grown asynchronously and relinked incrementally.

Something that might work: perhaps constructing Thrift() with just an `fs` argument creates all of the shared memo tables, as it does today, but doesn’t require an entry point or express permission to use the filesystem. Passing the filesystem is an “object capability”. Permission to use it is implied by having been given a reference to it.

From there, perhaps we expose an asynchronous `load` function that accepts an entryPoint calls back with the corresponding `Thrift` model instance.

If this only needs to work in modern JavaScript, I would memoize `load` on a shared table of module paths to promises for the corresponding Thrift object that resolve when the module is loaded, its transitive dependencies have also loaded, and the entire working set has been linked. If it needs to work on Node.js 0.10 (unlikely), the same can be accomplished with the usual ceremony involving `loaded` and `loading` state maps and lists of subscribers to the `loaded` event.

I believe that the link operation is already idempotent so it can safely be applied to incrementally link new modules as they are found.

I hope this makes sense to you.

kriskowal commented:
I should clarify, this all hinges on the fact that Thrift modules, just like JavaScript modules, can be loaded asynchronously but must be linked (executed) synchronously. You can read and compile a Thrift file just to read its `include` directives. They won’t be linkable until the module’s transitive dependencies have all loaded.

dbousque commented:
Hi @kriskowal, thanks for the additional details. Hopefully you had a nice time away :) Sorry for the late work on this, I've been a bit busy lately.

I added the ability to load sources asynchronously. It works by adding an AST building step before compilation. It was way easier to do it this way instead of modifying the compilation step to be able to be asynchronous, as I wanted to maintain backwards compatibility, and adding promises without disturbing the guarantee to have a fully built model after the constructor is done running was a bit tricky. In fact, I had to do some (a few lines worth of code) code duplication for both the asynchronous and the synchronous case. If it's ok to do some breaking changes, we could have a cleaner internal implementation if we could let go of the guarantee to get a fully built model when the constructor is done running (in the synchronous case).

Some changes that were done :
- Passing an entryPoint is no longer necessary, one can use the `thrift.load()` method to start the actual compilation.
- Allow to pass an object with a `load` field as the `fs` option parameter to the main `Thrift()` constructor : if passed, the `load` function will be used to fetch sources and will be expected to return a promise.
- Drop support for node 0.10 and start support from 0.12, because of the fact that native promises are missing. We can use a third-party promise library if necessary.
- The tests were modified to run twice : once with synchronous source loading, once with asynchronous source loading.

I don't know if you know about it, but there is an option to ignore whitespace changes when reviewing on GitHub, it can be handy in this case as most changes are whitespace changes.

kriskowal commented on thrift.js:
```diff
@@ -117,37 +118,112 @@ function Thrift(options) {
+    return Promise.resolve();
+}
+
+Thrift.prototype._buildChildAsts = function _buildChildAsts(filename, syntax, allowIncludeAlias, alreadyBuilt) {
```
nit: “build AST” is a long way to say “parse”.

Perhaps `_parseIncludedModules`?

kriskowal commented on thrift.js:
```diff
@@ -117,37 +118,112 @@ function Thrift(options) {
+    return this;
+}
+
+Thrift.prototype._buildAst = function _buildAst(filename, allowIncludeAlias, alreadyBuilt) {
```
`allowIncludeAlias` and `alreadyBuilt` should not need to be passed as arguments since they’re shared state of the transitive closure of `Thrift` objects.

Perhaps we should have a `this.parsed` map that we share among all `Thrift` objects?

Would that be redundant with presence in the `this.asts` map? It seems to me that being non-null there should be synonymous with having been parsed.

I do fully expect that we need `loaded` and `loading` maps. In that past, I’ve needed to distinguish the two to prevent duplicate effort.

dbousque replied:
Thanks again for your feedback. In the process of preparing the required changes, I would like to clarify something. `allowIncludeAlias` and `alreadyBuilt` were passed as argument because the included sources were fetched before any other `Thrift` object creation. We could share the `alreadyBuilt` map, but the `allowIncludeAlias` argument has a different purpose here (it is set to always `true` from the root's children on).

Because parsing is done before compiling, we don't build any other `Thrift` model in the potentially asynchronous part of the code. This piece of code protects from fetching a source twice, assuming it's atomic :
```javascript
if (alreadyBuilt[filename]) {
  return ;
}
alreadyBuilt[filename] = true;
```

Is this safe enough according to you, or the possibility for the piece of code above to be executed twice with the same `filename` argument at the same time should be accounted for?

dbousque replied:
Actually, is it even possible that the code above is executed twice at the same time? I don't know JavaScript very well (be it the spec or the implementations), but if I understand correctly, JavaScript engines won't run a "thread" until the one currently being executed is fully done executing. In our case, that means that the piece of code I shared is enough to guarantee that an async fetch could happen only once per `filename`, as there can't be a context switch between the `if` condition and `alreadyBuilt[filename] = true`.

Is that a correct interpretation? If not, could you expand a little on the `loaded` and `loading` maps you mentioned? I'm not familiar with these.

dbousque commented:
Hi @kriskowal. I've brought back support for 0.10, removed the use of promises and used callbacks instead, and followed your naming recommendations. I've split the PR in 4, so as to make reviewing easier, here they are :
- 1. Browser support : https://github.com/thriftrw/thriftrw-node/pull/165 (it is the PR we are currently speaking on).
- 2. Support for asynchronous source loading : https://github.com/dbousque/thriftrw-node/pull/1
- 3. Tests for 2. : https://github.com/dbousque/thriftrw-node/pull/2
- 4. Licence bump : https://github.com/dbousque/thriftrw-node/pull/3

Have a nice day.

kriskowal commented:
Thanks @dbousque. I’ve looped in engineers for review from the two areas where this code runs in production. I’ll try to find some time to spare to write a close review this week.

dbousque commented:
Thanks a lot @kriskowal.

kriskowal commented:
Approval contingent on test and coverage passing when the full stack is merged into this change.

dbousque commented:
Hi @kriskowal, I don't know if you saw, I addressed the issues raised in https://github.com/dbousque/thriftrw-node/pull/1. I'd understand if it's a bit much to review. For my purposes, browser support without asynchronous source loading is enough. Maybe we can merge this PR at least?

kriskowal commented:
I’ve landed this change, please go ahead and introduce the next change in your stack.

dbousque commented:
Neat! I created the PR for the asynchronous source loading feature : https://github.com/thriftrw/thriftrw-node/pull/168. You can see old comments on this on the old PR : https://github.com/dbousque/thriftrw-node/pull/1. It's expected that the tests don't pass, as there is a separate tests PR : https://github.com/dbousque/thriftrw-node/pull/2. I can merge both into a single PR if you'd prefer.

dbousque commented:
The PR has been adjusted since your latests comments. Let me know if you'd like the async source loading PR and the tests PR to be merged into a single one.
